### DISCUSSION TITLE
Fastest way to push large pd.DataFrame objects to Postgres table

### DISCUSSION THREAD

**Discussion by job-almekinders** [psycopg/psycopg#842] [General]
Hello,

First of all, thanks for maintaining this library! 

I'm trying to upload large pd.DataFrame's as efficiently as possible to a postgres table using psycopg3.

Currently, I'm writing to a postgres table using a io.StringIO object, like this: 
```python
csv_buffer = io.StringIO()
df.to_csv(csv_buffer, index=False)
csv_buffer.seek(0)

with cur.copy(f"COPY {table_name} FROM STDIN WITH CSV HEADER") as copy:
    copy.write(csv_buffer.getvalue())

conn.commit()
```

However, this loads the whole DataFrame into memory, which can become a problem if the DataFrame is really large.

Therefore, I would like to update this to use a spooled temp file (as this is recommended by @dvarrazzo in for example [this issue](https://github.com/psycopg/psycopg/issues/316)).

My currently implementation looks like this: 

```python
max_size = 1_000_000_000 # 1 GB

with SpooledTemporaryFile(max_size=max_size) as tmpfile:
    df.to_csv(tmpfile, index=False)
    tmpfile.seek(0)

    with cur.copy(f"COPY {table_name} FROM STDIN WITH CSV HEADER") as copy:
        for data in tmpfile:
            copy.write(data)

            conn.commit()
```

Would you think this is the most efficient way of doing this? 

Thanks in advance!

**Comment by dvarrazzo** [comment]
Hello,

there aren't so many possibilities: either you copy csv or you copy record by record, I guess.

So I propose that you run some benchmark yourself and then you report your results.
